### DISCUSSION TITLE
How to have a payload associated with each vector?

### DISCUSSION THREAD

**Discussion by habedi** [lmcinnes/pynndescent#203] [Q&A] [answered]
Hi!

I want to build an index of a database of high-dimensional numerical vectors, and do an approximate KNN search over the vectors. Each vector has a payload associated with it. By payload, I mean extra information such as timestamps, unique ID of each vector, etc. Now, I'm not sure how it should be done. It seems to me that ```pynndescent.NNDescent``` only uses the vectors as its first argument.

Any suggestion on how to add a payload to each vector while building an index?

BTW, thank you for the great work. An approximate KNN search library in Python saves many lives. :)

**Comment by lmcinnes** [comment]
If you simply want to reference the extra "payload" information after
search then I would just use a pandas dataframe and use the ``.iloc``
method to access the relevant information based on the nearest vector
indices. If you want the "payload" information used in the actual distance
computations that determine nearest points -- that's a lot more
complicated. You would probably have to write a custom distance metric and
do some careful data handling to be able to input it to pynndescent.

**Reply by habedi** [reply]
Sorry, I don't understand your answer so I write my question in a presumably more clear and concise way below.

What I want is something like this:

Suppose we have a set of records ```V``` and a record (belonging to ```V```) has the form ```{"vector": vector, "data": data}``` where ```vector``` is a high-dimensional vector, (for example, embedding of a word within a document) and the ```data``` is some arbitrary data (for example, the actual term which its embedding is stored in ```vector``` subfield).

Suppose I build an index on records in ```V``` using the information in ```vector``` subfield of records. 

Now, my question is, is it possible when I search for k nearest neighbours of a vector, I could get a reference to the information stored in the ```data``` subfield of the vectors that are returned in the result?

I'm assuming vectors are unique. So, if two records share identical information in their ```vector``` subfield, only the information for ```data``` of the second (or more recently inserted into the index) record is available.

**Comment by lmcinnes** [comment, marked as answer]
I think you want something like:

```python
# Split the data into raw vetcor data and other_data keeping order
vectors = np.asarray([v["vector"] for v in V])
other_data = [v["data"] for v in V]

# build an index over the vectors
index = NNDescent(vectors)
index.prepare()

...

# we can query a vector and get the index and distances of the nearest neighbors
neighbors, neighbor_distances = index.query([query_vector])
# and then find the appropriate other_data for each neighbor in the returned data
neighbor_data = [other_data[i] for i in neighbors[0]]
```

So the answer is yes, you can, but it isn't going to be automatic -- you'll have to write a small amount of extra code to convert the information as required.

**Reply by habedi** [reply]
Thank you for the code and clarification. Your answer solved my problem. As I understand it now, the result of a search returns the list of indices of neighbours of the query vectors in the input sequence (for example, the list of vectors V). And one could use the indices to reference other information, such as a data payload, that are associated with each vector.
